core/src/telemetry/log_export.rs

```diff
@@ -93,10 +96,78 @@ where
             fields,
             span_contexts: spans,
         };
+        self.consumer.on_log(log);
+    }
+}
+
+/// Core log consumer implementation backed by a ring buffer.
+pub struct CoreLogBufferedConsumer {
+    logs_in: Mutex<Producer<CoreLog, Arc<HeapRb<CoreLog>>>>,
+}
+
+impl CoreLogBufferedConsumer {
+    /// Create a log consumer and drainable buffer for the given capacity.
+    pub fn new(ringbuf_capacity: usize) -> (Self, CoreLogBuffer) {
+        let (logs_in, logs_out) = HeapRb::new(ringbuf_capacity).split();
+        (
+            Self {
+                logs_in: Mutex::new(logs_in),
+            },
+            CoreLogBuffer { logs_out },
+        )
+    }
+}
+
+impl CoreLogConsumer for CoreLogBufferedConsumer {
+    fn on_log(&self, log: CoreLog) {
         let _ = self.logs_in.lock().push(log);
     }
 }
 
+impl fmt::Debug for CoreLogBufferedConsumer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("<buffered consumer>")
+    }
+}
+
+/// Buffer of core logs that can be drained.
+pub struct CoreLogBuffer {
+    logs_out: CoreLogsOut,
+}
+
+impl CoreLogBuffer {
+    /// Drain the buffer of its logs.
+    pub fn drain(&mut self) -> Vec<CoreLog> {
+        self.logs_out.pop_iter().collect()
+    }
+}
+
+/// Core log consumer implementation backed by a mpsc channel.
+pub struct CoreLogStreamConsumer {
+    tx: Sender<CoreLog>,
+}
+
+impl CoreLogStreamConsumer {
+    /// Create a stream consumer and stream of logs.
+    pub fn new(buffer: usize) -> (Self, Receiver<CoreLog>) {
```

Comment:
It is clearer here to return `Receiver` than some kind of `Box<dyn Stream<Item = CoreLog>>` because it's not sized which is required by some `StreamExt` calls

Reply:
You can just return `impl Stream<Item = CoreLog>>`, but this is fine too.

Reply:
:+1: I wasn't able to cleanly mark the trait as sized which was biting me when using some of the `StreamExt` calls (though can uncleanly do it if I must). Yeah, I usually wouldn't want to leak the specific trait impl I am using to the caller, but that it's sized and this is not some user-facing API settled it for me.